### DISCUSSION TITLE
Publishing messages from the HTTP API

### DISCUSSION THREAD

**Discussion by AmirDotDev0** [emqx/emqx#8510] [Q&A] [answered]
Hello everyone, I'm creating an android chat app based on emqx, the app publishes messages through the HTTP API instead of the client (paho android client), what I need to know is: Will publishing messages through the HTTP API affect the server performance? or emqx performance?

**Comment by zmstone** [comment, marked as answer]
Hi @AmirOsman2020 

* HTTP in general is a heavier protocol than MQTT.
* Internally in EMQX, HTTP request parsing has more overheads than MQTT messages, however unlikely an order of magnitude difference.
* If HTTP connection is closed for each request, then the overhead of connect/reconnect should be considered, especially when using TLS.

I am not aware if anyone has done performance tests for the HTTP API before, it would be interesting to know.

**Reply by AmirDotDev0** [reply]
Ok, am not familiar with the way the HTTP API works internally, but according to your answer am assuming that in the case of the publish HTTP request "http://localhost:8081/api/v4/mqtt/publish" an MQTT connection will be established, the message will be published and it will disconnect, I oversimplified but is this correct?
In case the above assumption is correct, you said "If HTTP connection is closed for each request" does that mean I can make the publish HTTP request "http://localhost:8081/api/v4/mqtt/publish" without closing the connection? if this is true how is it possible to make publish request without closing the connection, as far as I know, HTTP/1 always closes the connection? can you provide more details?

Note:
Please correct me if I'm wrong, when you said "If HTTP connection is closed for each request" I assumed that you are talking about a connection to the emqx broker eternally , did you mean the mqtt connection eternally or just the HTTP request connection?

**Reply by zmstone** [reply]
* There is http pipelining, it’s supported by the http server library EMQX is using, however I am not sure if it works because I have never tested it, that’s why I used “if”
* the `/mqtt/publish_batch` endpoint can save some overheads of per message request/response.

**Reply by AmirDotDev0** [reply]
Ok, noted, thanks for the help.
